### DISCUSSION TITLE
Simple check() function without Pydantic models

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [pydantic/pydantic#6039] [Pydantic V2]
I'd really like to have a simple interface that works with standard typing annotations and the minimum amount of Pydantic-ness possible. I'm wondering if this would be of interest (in which case I'm happy to help work on this)?

The rough API I'd like to see is something like:
```
from pydantic import check, Validators

def convert_to_int(value: str) -> int:
    return {"foo": 4, "bar": 5}.get(value, 1)

def divide_int(value: int) -> float:
    return value / 3

class MyDict(TypedDict):
    a: Annotated[float, Validators[str.lower, convert_to_int, divide_int]]
    b: int
    c: MyObj

result = check(MyDict, {"a": "Foo", "b": "42", "c": MyObj()})  # -> {"a": 1.33333, "b": 42, "c": <__main__.MyObj object at 0x7f2b05de7dc0>}
```

Where the input type of `a` is taken from the input type of the first validator function. A mypy plugin could also be used to check that all annotations chain together correctly (i.e. str.lower returns str, so convert_to_int must accept str, divide_int must accept int and return float for the final value).

So far, I think the `check()` function can be created with something like:
```
@lru_cache
def _get_schema(t):
    return SchemaValidator(GenerateSchema(ConfigWrapper({"arbitrary_types_allowed": True}), None).generate_schema(t))

def check(t, value):
    return _get_schema(t).validate_python(value)
```

I'm not so sure how to do the validators currently. I can see that `PlainValidator` can be used to wrap each function, but when there are multiple of them, it only applies the last function. So, I guess I'd need to create a `Validators` object which gets converted into a sequence of `PlainValidator`s or similar.

**Comment by adriangb** [comment]
Would `check = lambda t, v: pydantic.TypeAdapter(t).validate_python(v)` work for you? Combined with `a: Annotated[str, pydantic.AfterValidator(lambda x: str.lower(x)), pydantic.AfterValidator(…)]`. Using a `BeforeValidator`, `AfterValidator` or `WrapValidator` you’ll be able to apply multiple validators that get chained.

**Reply by Dreamsorcerer** [reply]
Using TypeAdapter doesn't allow the config to be changed on some types (disallowing arbitrary types would seem rather weird for something that uses straight annotations). I think those validators also would not allow an arbitrary number of functions, so I was thinking about a new class which accepts multiple functions (and again, having a mypy plugin to improve typing would be great).

Also, my initial example above uses lru_cache() to avoid rebuilding the schema on every call.

My question is just would something like this be of interest to be included in Pydantic? If yes, we can discuss implementation details later.

**Reply by adriangb** [reply]
> Using TypeAdapter doesn't allow the config to be changed on some types (disallowing arbitrary types would seem rather weird for something that uses straight annotations).

Allowing the config to be changed for existing types is a whole can of worms. It's something we're aware of and thinking of, but it's not simple. cc @dmontagu since he's interested in this as well.

> I think those validators also would not allow an arbitrary number of functions, so I was thinking about a new class that accepts multiple functions (and again, having a mypy plugin to improve typing would be great).

Not in a single call but like I showed above you can have multiple validators each with a function.

> I was thinking about a new class which accepts multiple functions (and again, having a mypy plugin to improve typing would be great).

That does sound really interesting and is something I've thought about! I wonder if a builder pattern like `Validator(str.lower).chain(int).chain(...)` might obviate the need to a mypy plugin. We're definitely open to new APIs like this if (1) they are request by several users and (2) after the 2.0 release.

> Also, my initial example above uses lru_cache() to avoid rebuilding the schema on every call.

Caching is also more complex than it seems and is something we cannot do at the Pydantic level: we've had use cases where people create thousands of ephemeral models in a loop and we used to leak memory because we had some caches that referenced them. Because of this we now avoid any sort of caching that might persist references to user's objects and prevent them from getting garbage collected. Besides, it isn't too had to add a cache around calls to create `TypeAdapter`s.

**Reply by Dreamsorcerer** [reply]
> Allowing the config to be changed for existing types is a whole can of worms. It's something we're aware of and thinking of, but it's not simple. cc @dmontagu since he's interested in this as well.

OK, seems to work without a problem from what I can see. Note that 'existing types' in this case basically means `TypedDict` or `dataclass`.

> Not in a single call but like I showed above you can have multiple validators each with a function.

But, only 2 or 3, right? I mentioned this at the end of the original question, but it seems that only the last validator of a given type is run. So, you can have a BeforeValidator and an AfterValidator, but not 2+ AfterValidators. Which makes it rather restrictive and awkward to figure out the right sequence of validators.

> > I was thinking about a new class which accepts multiple functions (and again, having a mypy plugin to improve typing would be great).
> 
> That does sound really interesting and is something I've thought about! I wonder if a builder pattern like `Validator(str.lower).chain(int).chain(...)` might obviate the need to a mypy plugin.

I'll have a think through that and see if I can make that work. I guess if the object is `Generic[InputType, OutputType]` then something like `def chain(v: Callable[[OutputType], NewOutputType]) -> Validator[InputType, NewOutputType]` might work. I think you'd still need a plugin to set the input type and validate the output type for the annotated type itself though.

Another thought is that you could have Pydantic validate types between each validator function (by default, maybe only when an output type is `Any` and the next input is not `Any`/`object`).

> > Also, my initial example above uses lru_cache() to avoid rebuilding the schema on every call.
> 
> Caching is also more complex than it seems and is something we cannot do at the Pydantic level: we've had use cases where people create thousands of ephemeral models in a loop and we used to leak memory because we had some caches that referenced them. Because of this we now avoid any sort of caching that might persist references to user's objects and prevent them from getting garbage collected. Besides, it isn't too had to add a cache around calls to create `TypeAdapter`s.

Hmm, I see. In this case, it doesn't seem so risky to me though. If a user is going to use this new function, they will probably use this across their application. lru_cache() is limited to 128 objects by default, so if lots of objects are created they wouldn't cause a memory issue, and it's a global function, so there won't be multiple caches or anything either.

**Reply by adriangb** [reply]
> But, only 2 or 3, right? I mentioned this at the end of the original question, but it seems that only the last validator of a given type is run. So, you can have a BeforeValidator and an AfterValidator, but not 2+ AfterValidators. Which makes it rather restrictive and awkward to figure out the right sequence of validators.

That only applies to `PlainValidator`:

```python
from typing import Annotated

from pydantic import AfterValidator, TypeAdapter

Foo = Annotated[int, AfterValidator(lambda x: x * 2), AfterValidator(lambda x: x + 1)]

assert TypeAdapter(Foo).validate_python(3) == 7
```

> Another thought is that you could have Pydantic validate types between each validator function (by default, maybe only when an output type is Any and the next input is not Any/object).

Yes in theory we could take the generated `CoreSchema` and do some sort of runtime type checking on them, but it gets hard with introspecting the signature of functions and whatnot. Definitely not something we're going to do _soon_.

**Reply by Dreamsorcerer** [reply]
> That only applies to `PlainValidator`:

Ah yes, I see now. If I change the type to the input type, then I can use multiple AfterValidators. If I use BeforeValidators, then they need to be reversed for some reason, they seem to evaluate right-to-left. I think these details tripped me up when I tested them originally.

**Reply by adriangb** [reply]
Understandable. It’s an extremely powerful but somewhat confusing system.
Better docs will help but please feel free to report any ideas/observations
you have.

**Reply by Dreamsorcerer** [reply]
@adriangb Any changes on this front yet? It still seems like I need to use the private SchemaValidator shown in my original question in order to allow arbitrary types.

**Reply by Dreamsorcerer** [reply]
In fact, that code doesn't even work now...
